src/main/java/org/fcrepo/storage/ocfl/DefaultOcflObjectSession.java

```diff
@@ -142,30 +147,48 @@ public synchronized void writeResource(final ResourceHeaders headers, final Inpu
         final var contentPath = encode(paths.getContentFilePath());
         final var headerPath = encode(paths.getHeaderFilePath());
 
+        final var contentDst = createStagingPath(contentPath);
+        final var headerDst = createStagingPath(headerPath);
+
         deletePaths.remove(contentPath);
         deletePaths.remove(headerPath);
 
-        if (content != null) {
-            final var contentDst = createStagingPath(contentPath);
-            var digest = getOcflDigest(headers.getDigests());
+        try {
+            if (content != null) {
+                var digest = getOcflDigest(headers.getDigests());
+
+                if (digest == null) {
+                    // compute the digest that OCFL uses if it was not provided
+                    final var messageDigest = digestAlgorithm.getMessageDigest();
+                    write(new DigestInputStream(content, messageDigest), contentDst);
+                    digest = Hex.encodeHexString(messageDigest.digest());
+                    addDigestHeader(digest, headers);
+                } else {
+                    write(content, contentDst);
+                }
 
-            if (digest == null) {
-                final var messageDigest = digestAlgorithm.getMessageDigest();
-                write(new DigestInputStream(content, messageDigest), contentDst);
-                digest = Hex.encodeHexString(messageDigest.digest());
-                addDigestHeader(digest, headers);
-            } else {
-                write(content, contentDst);
-            }
+                digests.put(contentPath, digest);
 
-            digests.put(contentPath, digest);
+                final var fileSize = fileSize(contentDst);
 
-            headers.setContentPath(contentPath.path);
-            headers.setContentSize(fileSize(contentDst));
-        }
+                if (headers.getContentSize() != null
+                        && fileSize != headers.getContentSize()) {
+                    throw new RuntimeException(String.format("Resource %s's file size does not match expectation." +
```

Comment:
I had considered creating a new exception, and will do so. I ultimately did not because the Fedora code currently treats content size mismatches as repository runtime exceptions, so the behavior would not change regardless of the exception type thrown here.